File: moveZeroToEnd.js
/*
* Given an array of integer value eg: [2,4,0,4,3,-1,0,5,0]
* Move all zero value to the end of array list
* expected: [2,4,4,3,-1,5,0,0,0]
*/

const solutionOne = (arr) => {
  for(let item in arr){
    if(arr[item] === 0){
      arr.splice(item, 1);
      arr.push(0);
    }
  }
  return arr;
}

console.log("solution 1: ", solutionOne([2, 4, 0, 4, 3, -1, 0, 5, 0]));


const solutionTwo = (arr) => {
  let m = 0, temp;

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] != 0) {
      temp = arr[i];
      arr[i] = arr[m];
      arr[m] = temp;
      m++;
    }
  }

  return arr;
};

console.log("solution 2: ", solutionTwo([2, 4, 0, 4, 3, -1, 0, 5, 0]));

const solutionThree = (arr) =>{
  let j = 0;
  let i = 1;
  while (i < arr.length) {
    if (arr[j] === 0 && arr[i] !== 0) {
      arr[j] = arr[i];
      arr[i] = 0;
      j++;
    }
    if (arr[j] !== 0) {
      j++;
    }
    i++;
  }

  return arr;
}

console.log("solution 3: ", solutionThree([2, 4, 0, 4, 3, -1, 0, 5, 0]));

File: python/moveZeroToEnd.py
# Solution One
# def solution_one(arr):
#     index = 0
#     while index < len(arr):
#         if arr[index] == 0:
#             arr.remove(0)
#             arr.append(0)
#         else:
#             index += 1
#     return arr

# print("Solution 1:", solution_one([2, 4, 0, 4, 3, -1, 0, 5, 0]))


# Solution Two
def solution_two(arr):
    m = 0
    for i in range(len(arr)):
        if arr[i] != 0:
            arr[i], arr[m] = arr[m], arr[i]
            m += 1
    return arr

print("Solution 2:", solution_two([2, 4, 0, 4, 3, -1, 0, 5, 0]))


# # Solution Three
def solution_three(arr):
    j = 0
    i = 1
    while i < len(arr):
        if arr[j] == 0 and arr[i] != 0:
            arr[j] = arr[i]
            arr[i] = 0
            j += 1
        if arr[j] != 0:
            j += 1
        i += 1
    return arr

print("Solution 3:", solution_three([2, 4, 0, 4, 3, -1, 0, 5, 0]))
